File: languages/c-plus-plus/cross-language-features.cpp
#include <iostream>
#include <utility>
#include "../../runtimes/c-plus-plus/willyhorizont/runtime/xl.hpp"

int main(int argc, char* argv[]) {
    /*
    1. support lambda as value, or has workaround
    */
    XL::Type say_hello = XL::init_lambda([](const XL::Type& va) -> XL::Type {
        XL::Type itr = va.iter();
        XL::Type callback = itr.next();
        std::cout << "hello" << '\n';
        callback.call();
        return XL::NONE;
    });
    say_hello.call(XL::init_lambda([](const XL::Type& va) -> XL::Type {
        std::cout << "world" << '\n';
        return XL::NONE;
    }));
    XL::Type create_multiplier = XL::init_lambda([](const XL::Type& va) -> XL::Type {
        XL::Type itr = va.iter();
        XL::Type aa = itr.next();
        return XL::init_lambda([aa = std::move(aa)](const XL::Type& va) -> XL::Type {
            XL::Type itr = va.iter();
            XL::Type bb = itr.next();
            return XL::init_int(XL::to_int(aa) * XL::to_int(bb));
        });
    });
    XL::Type multiply_by_two = create_multiplier.call(XL::init_int(2));
    std::cout << "multiply_by_two(10): " << XL::to_int(multiply_by_two.call(XL::init_int(10))) << '\n';
    XL::Type multiply_by_eight = create_multiplier.call(XL::init_int(8));
    std::cout << "multiply_by_eight(4): " << XL::to_int(multiply_by_eight.call(XL::init_int(4))) << '\n';
    std::cout << "multiply_by_two(8): " << XL::to_int(multiply_by_two.call(XL::init_int(8))) << '\n';

    /*
    2. support dynamic-typed value, or has workaround
    */
    XL::Type xl_list = XL::init_list(
        XL::NONE,
        XL::TRUE,
        XL::FALSE,
        XL::init_string("foo"),
        XL::init_int(0),
        XL::init_int(-123),
        XL::init_float(123.789),
        XL::init_float(-123.789),
        XL::init_list(XL::init_int(1), XL::init_int(2), XL::init_int(3)),
        XL::init_dict(XL::Pair("foo", XL::init_string("bar"))),
        XL::init_lambda([](const XL::Type& va) -> XL::Type {
            XL::Type itr = va.iter();
            XL::Type aa = itr.next();
            XL::Type bb = itr.next();
            return XL::init_int(XL::to_int(aa) * XL::to_int(bb));
        })
    );
    std::cout << "xl_list: " << XL::json_stringify(xl_list) << '\n';
    std::cout << "xl_list: " << XL::json_stringify(xl_list, { .pretty = true }) << '\n';
    XL::Type xl_dict = XL::init_dict(
        XL::Pair("xl_none", XL::NONE),
        XL::Pair("xl_bool_true", XL::TRUE),
        XL::Pair("xl_bool_false", XL::FALSE),
        XL::Pair("xl_string", XL::init_string("foo")),
        XL::Pair("xl_int_positive", XL::init_int(0)),
        XL::Pair("xl_int_negative", XL::init_int(-123)),
        XL::Pair("xl_float_positive", XL::init_float(123.789)),
        XL::Pair("xl_float_negative", XL::init_float(-123.789)),
        XL::Pair("xl_list", XL::init_list(XL::init_int(1), XL::init_int(2), XL::init_int(3))),
        XL::Pair("xl_dict", XL::init_dict(XL::Pair("foo", XL::init_string("bar")))),
        XL::Pair("xl_lambda", XL::init_lambda([](const XL::Type& va) -> XL::Type {
            XL::Type itr = va.iter();
            XL::Type aa = itr.next();
            XL::Type bb = itr.next();
            return XL::init_int(XL::to_int(aa) * XL::to_int(bb));
        }))
    );
    std::cout << "xl_dict: " << XL::json_stringify(xl_dict) << '\n';
    std::cout << "xl_dict: " << XL::json_stringify(xl_dict, { .pretty = true }) << '\n';

    return 0;
}

File: languages/objective-c/cross-language-features.m
#import <Foundation/Foundation.h>
#import "../../runtimes/objective-c/willyhorizont/runtime/xl.h"

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        XL * sayHello = xl.initClosure(^(XL * args) {
            XL * callbackFunction = [args next];
            NSLog(@"hello");
            [callbackFunction call:@[]];
            return xl.initNone();
        });
        [sayHello call:@[xl.initClosure(^(XL * a) {
            NSLog(@"world");
            return xl.initNone();
        })]];
        XL * createMultiplier = xl.initClosure(^(XL * args) {
            XL * aa = [args next];
            return xl.initClosure(^(XL * args) {
                XL * bb = [args next];
                return xl.initInt(xl.toInt(aa) * xl.toInt(bb));
            });
        });
        XL * multiplyByTwo = [createMultiplier call:@[xl.initInt(2)]];
        NSLog(@"%@", [NSString stringWithFormat:@"multiply_by_two(10): %@", xl.jsonStringify([multiplyByTwo call:@[xl.initInt(10)]], nil)]);
        XL * multiplyByEight = [createMultiplier call:@[xl.initInt(8)]];
        NSLog(@"%@", [NSString stringWithFormat:@"multiply_by_eight(4): %@", xl.jsonStringify([multiplyByEight call:@[xl.initInt(4)]], nil)]);
        NSLog(@"%@", [NSString stringWithFormat:@"multiply_by_two(2): %@", xl.jsonStringify([multiplyByTwo call:@[xl.initInt(2)]], nil)]);
        XL * xlList = xl.initList(@[
            xl.initNone(),
            xl.initBool(YES),
            xl.initBool(NO),
            xl.initString(@"foo"),
            xl.initInt(0),
            xl.initInt(-123),
            xl.initFloat(123.789),
            xl.initFloat(-123.789),
            xl.initList(@[xl.initInt(1), xl.initInt(2), xl.initInt(3)]),
            xl.initDict(@{ @"foo": xl.initString(@"bar") }),
            xl.initClosure(^(XL * args) {
                XL * aa = [args next];
                XL * bb = [args next];
                return xl.initInt(xl.toInt(aa) * xl.toInt(bb));
            }),
        ]);
        NSLog(@"%@", [NSString stringWithFormat:@"xl_list: %@", xl.jsonStringify(xlList, nil)]);
        NSLog(@"%@", [NSString stringWithFormat:@"xl_list: %@", xl.jsonStringify(xlList, xl.initDict(@{ @"pretty": xl.initBool(YES) }))]);
        XL * xlDict = xl.initDict(@{
            @"xl_none": xl.initNone(),
            @"xl_bool_true": xl.initBool(YES),
            @"xl_bool_false": xl.initBool(NO),
            @"xl_string": xl.initString(@"foo"),
            @"xl_int_positive": xl.initInt(0),
            @"xl_int_negative": xl.initInt(-123),
            @"xl_float_positive": xl.initFloat(123.789),
            @"xl_float_negative": xl.initFloat(-123.789),
            @"xl_list": xl.initList(@[xl.initInt(1), xl.initInt(2), xl.initInt(3)]),
            @"xl_dict": xl.initDict(@{ @"foo": xl.initString(@"bar") }),
            @"xl_closure": xl.initClosure(^(XL * args) {
                XL * aa = [args next];
                XL * bb = [args next];
                return xl.initInt(xl.toInt(aa) * xl.toInt(bb));
            }),
        });
        NSLog(@"%@", [NSString stringWithFormat:@"xl_dict: %@", xl.jsonStringify(xlDict, nil)]);
        NSLog(@"%@", [NSString stringWithFormat:@"xl_dict: %@", xl.jsonStringify(xlDict, xl.initDict(@{ @"pretty": xl.initBool(YES) }))]);
    }
    return 0;
}
